File: patterns/ai/provider-gateway/files/ts/src/ai/gateway.ts
/**
 * One call signature over every provider.
 *
 * The point is NOT to abstract providers away — provider-specific features are
 * why you pick a provider. The point is that the boring 90% (auth, retries,
 * usage accounting, timeouts, fallback, budget) is written once, and the
 * interesting 10% stays reachable through `raw`.
 *
 * Adding a provider means adding one adapter below and one entry in
 * config/models.json. Nothing else in your codebase changes.
 */
import { env } from "../config/env.js";
import { costOf, estimateMaxCost, fallbacksFor, resolveModel, spec } from "./models.js";

export interface Message {
  role: "system" | "user" | "assistant";
  content: string;
}

export interface CompleteOptions {
  /** Alias ("fast" | "default" | "deep") or a concrete model id. */
  model?: string;
  messages: Message[];
  maxOutputTokens?: number;
  temperature?: number;
  /** Abort the call after this many ms. Always set one in production. */
  timeoutMs?: number;
  /** Retries on transient failures, per model tried. */
  maxRetries?: number;
  /** Try the registry's fallback models when the primary keeps failing. */
  allowFallback?: boolean;
  /** Hard ceiling for this call. Defaults to AI_MAX_USD_PER_RUN. */
  maxUsd?: number;
  signal?: AbortSignal;
}

export interface Completion {
  text: string;
  model: string;
  usage: { inputTokens: number; outputTokens: number; cachedInputTokens?: number };
  costUsd: number;
  /** Models tried and rejected before this one succeeded. */
  attempted: string[];
  /** The untouched provider response, for anything this interface flattens. */
  raw: unknown;
}

export class BudgetExceededError extends Error {}
export class MissingCredentialError extends Error {}

/** Errors worth retrying: transient by nature, not caused by the request. */
const RETRYABLE_STATUS = new Set([408, 409, 429, 500, 502, 503, 504]);

interface ProviderError extends Error {
  status?: number;
}

function isRetryable(err: unknown): boolean {
  const e = err as ProviderError;
  if (e?.name === "AbortError") return false;
  if (typeof e?.status === "number") return RETRYABLE_STATUS.has(e.status);
  // Network-level failures have no status and are worth one more try.
  return e instanceof TypeError;
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

export async function complete(opts: CompleteOptions): Promise<Completion> {
  const primary = resolveModel(opts.model ?? "default");
  const chain = opts.allowFallback === false ? [primary] : [primary, ...fallbacksFor(primary)];
  const attempted: string[] = [];

  let lastError: unknown;
  for (const modelId of chain) {
    attempted.push(modelId);
    try {
      return { ...(await callWithRetries(modelId, opts)), attempted };
    } catch (err) {
      // A budget or credential problem will not be fixed by another model.
      if (err instanceof BudgetExceededError || err instanceof MissingCredentialError) throw err;
      if (!isRetryable(err)) throw err;
      lastError = err;
    }
  }
  throw lastError;
}

async function callWithRetries(
  modelId: string,
  opts: CompleteOptions,
): Promise<Omit<Completion, "attempted">> {
  const maxRetries = opts.maxRetries ?? 2;
  const budget = opts.maxUsd ?? env.AI_MAX_USD_PER_RUN;

  // Refuse before spending, not after. A cheap approximation of input size is
  // enough here — being wrong by 20% does not matter when the ceiling exists
  // to catch a runaway loop, not to bill accurately.
  const approxInputTokens = Math.ceil(
    opts.messages.reduce((n, m) => n + m.content.length, 0) / 4,
  );
  const worstCase = estimateMaxCost(modelId, approxInputTokens, opts.maxOutputTokens);
  if (worstCase > budget) {
    throw new BudgetExceededError(
      `${modelId} could cost up to $${worstCase.toFixed(2)}, over the $${budget.toFixed(2)} ceiling. ` +
        `Raise maxUsd / AI_MAX_USD_PER_RUN, cap maxOutputTokens, or use a cheaper alias.`,
    );
  }

  let lastError: unknown;
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await callProvider(modelId, opts);
    } catch (err) {
      lastError = err;
      if (attempt === maxRetries || !isRetryable(err)) throw err;
      // Exponential backoff with jitter: without jitter, a fleet of workers
      // retries in lockstep and re-creates the spike that rate-limited them.
      const base = 500 * 2 ** attempt;
      await sleep(base + Math.random() * base);
    }
  }
  throw lastError;
}

async function callProvider(
  modelId: string,
  opts: CompleteOptions,
): Promise<Omit<Completion, "attempted">> {
  const provider = spec(modelId).provider;
  const signal = opts.signal ?? AbortSignal.timeout(opts.timeoutMs ?? 60_000);

  switch (provider) {
    case "anthropic":
      return anthropic(modelId, opts, signal);
    case "openai":
      return openai(modelId, opts, signal);
    case "google":
      return google(modelId, opts, signal);
    default: {
      const exhaustive: never = provider;
      throw new Error(`No adapter for provider '${exhaustive}'`);
    }
  }
}

function requireKey(value: string | undefined, name: string, provider: string): string {
  if (!value) {
    throw new MissingCredentialError(
      `${provider} needs ${name}. Add it to .env (see .env.example) or pick a model from another provider.`,
    );
  }
  return value;
}

async function post(url: string, init: RequestInit): Promise<unknown> {
  const res = await fetch(url, init);
  if (!res.ok) {
    const err: ProviderError = new Error(`${res.status} ${res.statusText}: ${await res.text()}`);
    err.status = res.status;
    throw err;
  }
  return res.json();
}

// ---- adapters ------------------------------------------------------------
// Each one does exactly two jobs: shape the request, and normalise the usage
// numbers. Everything else is handled above.

async function anthropic(modelId: string, o: CompleteOptions, signal: AbortSignal) {
  const key = requireKey(env.ANTHROPIC_API_KEY, "ANTHROPIC_API_KEY", "Anthropic");
  const system = o.messages.filter((m) => m.role === "system").map((m) => m.content).join("\n\n");
  const raw = (await post("https://api.anthropic.com/v1/messages", {
    method: "POST",
    signal,
    headers: {
      "content-type": "application/json",
      "x-api-key": key,
      "anthropic-version": "2023-06-01",
    },
    body: JSON.stringify({
      model: modelId,
      max_tokens: o.maxOutputTokens ?? 4096,
      ...(o.temperature !== undefined && { temperature: o.temperature }),
      ...(system && { system }),
      messages: o.messages.filter((m) => m.role !== "system"),
    }),
  })) as {
    content: Array<{ type: string; text?: string }>;
    usage: {
      input_tokens: number;
      output_tokens: number;
      cache_read_input_tokens?: number;
    };
  };

  const usage = {
    inputTokens: raw.usage.input_tokens,
    outputTokens: raw.usage.output_tokens,
    cachedInputTokens: raw.usage.cache_read_input_tokens,
  };
  return {
    text: raw.content.filter((b) => b.type === "text").map((b) => b.text ?? "").join(""),
    model: modelId,
    usage,
    costUsd: costOf(modelId, usage),
    raw,
  };
}

async function openai(modelId: string, o: CompleteOptions, signal: AbortSignal) {
  const key = requireKey(env.OPENAI_API_KEY, "OPENAI_API_KEY", "OpenAI");
  const raw = (await post("https://api.openai.com/v1/chat/completions", {
    method: "POST",
    signal,
    headers: { "content-type": "application/json", authorization: `Bearer ${key}` },
    body: JSON.stringify({
      model: modelId,
      messages: o.messages,
      ...(o.maxOutputTokens && { max_completion_tokens: o.maxOutputTokens }),
      ...(o.temperature !== undefined && { temperature: o.temperature }),
    }),
  })) as {
    choices: Array<{ message: { content: string | null } }>;
    usage: {
      prompt_tokens: number;
      completion_tokens: number;
      prompt_tokens_details?: { cached_tokens?: number };
    };
  };

  const usage = {
    inputTokens: raw.usage.prompt_tokens,
    outputTokens: raw.usage.completion_tokens,
    cachedInputTokens: raw.usage.prompt_tokens_details?.cached_tokens,
  };
  return {
    text: raw.choices[0]?.message.content ?? "",
    model: modelId,
    usage,
    costUsd: costOf(modelId, usage),
    raw,
  };
}

async function google(modelId: string, o: CompleteOptions, signal: AbortSignal) {
  const key = requireKey(env.GOOGLE_API_KEY, "GOOGLE_API_KEY", "Google");
  const system = o.messages.filter((m) => m.role === "system").map((m) => m.content).join("\n\n");
  const raw = (await post(
    `https://generativelanguage.googleapis.com/v1beta/models/${modelId}:generateContent`,
    {
      method: "POST",
      signal,
      headers: { "content-type": "application/json", "x-goog-api-key": key },
      body: JSON.stringify({
        ...(system && { systemInstruction: { parts: [{ text: system }] } }),
        contents: o.messages
          .filter((m) => m.role !== "system")
          .map((m) => ({
            role: m.role === "assistant" ? "model" : "user",
            parts: [{ text: m.content }],
          })),
        generationConfig: {
          ...(o.maxOutputTokens && { maxOutputTokens: o.maxOutputTokens }),
          ...(o.temperature !== undefined && { temperature: o.temperature }),
        },
      }),
    },
  )) as {
    candidates?: Array<{ content?: { parts?: Array<{ text?: string }> } }>;
    usageMetadata: {
      promptTokenCount: number;
      candidatesTokenCount: number;
      cachedContentTokenCount?: number;
    };
  };

  const usage = {
    inputTokens: raw.usageMetadata.promptTokenCount,
    outputTokens: raw.usageMetadata.candidatesTokenCount,
    cachedInputTokens: raw.usageMetadata.cachedContentTokenCount,
  };
  return {
    text: raw.candidates?.[0]?.content?.parts?.map((p) => p.text ?? "").join("") ?? "",
    model: modelId,
    usage,
    costUsd: costOf(modelId, usage),
    raw,
  };
}

File: patterns/ai/provider-gateway/files/py/src/ai/gateway.py
"""One call signature over every provider.

The point is NOT to abstract providers away — provider-specific features are
why you pick a provider. The point is that the boring 90% (auth, retries,
usage accounting, timeouts, fallback, budget) is written once, and the
interesting 10% stays reachable through ``Completion.raw``.

Adding a provider means adding one adapter below and one entry in
``config/models.json``. Nothing else in your codebase changes.
"""

from __future__ import annotations

import random
import time
from dataclasses import dataclass, field
from typing import Any, Literal

import httpx

from ..config.env import settings
from .models import cost_of, estimate_max_cost, fallbacks_for, resolve_model, spec

Role = Literal["system", "user", "assistant"]

RETRYABLE_STATUS = {408, 409, 429, 500, 502, 503, 504}


@dataclass
class Message:
    role: Role
    content: str


@dataclass
class Usage:
    input_tokens: int
    output_tokens: int
    cached_input_tokens: int = 0


@dataclass
class Completion:
    text: str
    model: str
    usage: Usage
    cost_usd: float
    attempted: list[str] = field(default_factory=list)
    raw: Any = None


class BudgetExceededError(RuntimeError):
    pass


class MissingCredentialError(RuntimeError):
    pass


def _is_retryable(exc: BaseException) -> bool:
    if isinstance(exc, httpx.HTTPStatusError):
        return exc.response.status_code in RETRYABLE_STATUS
    # Network-level failures carry no status and are worth another try.
    return isinstance(exc, (httpx.TimeoutException, httpx.TransportError))


def complete(
    messages: list[Message],
    *,
    model: str = "default",
    max_output_tokens: int | None = None,
    temperature: float | None = None,
    timeout_s: float = 60.0,
    max_retries: int = 2,
    allow_fallback: bool = True,
    max_usd: float | None = None,
) -> Completion:
    primary = resolve_model(model)
    chain = [primary] + (fallbacks_for(primary) if allow_fallback else [])
    attempted: list[str] = []
    last_exc: BaseException | None = None

    for model_id in chain:
        attempted.append(model_id)
        try:
            result = _call_with_retries(
                model_id,
                messages,
                max_output_tokens=max_output_tokens,
                temperature=temperature,
                timeout_s=timeout_s,
                max_retries=max_retries,
                max_usd=max_usd,
            )
            result.attempted = attempted
            return result
        except (BudgetExceededError, MissingCredentialError):
            # Another model will not fix a budget or credential problem.
            raise
        except Exception as exc:  # noqa: BLE001 — re-raised below if not retryable
            if not _is_retryable(exc):
                raise
            last_exc = exc

    assert last_exc is not None
    raise last_exc


def _call_with_retries(
    model_id: str,
    messages: list[Message],
    *,
    max_output_tokens: int | None,
    temperature: float | None,
    timeout_s: float,
    max_retries: int,
    max_usd: float | None,
) -> Completion:
    budget = max_usd if max_usd is not None else settings.ai_max_usd_per_run

    # Refuse before spending, not after. A cheap approximation of input size is
    # enough — being wrong by 20% does not matter when the ceiling exists to
    # catch a runaway loop, not to bill accurately.
    approx_input = sum(len(m.content) for m in messages) // 4
    worst_case = estimate_max_cost(model_id, approx_input, max_output_tokens)
    if worst_case > budget:
        raise BudgetExceededError(
            f"{model_id} could cost up to ${worst_case:.2f}, over the "
            f"${budget:.2f} ceiling. Raise max_usd / AI_MAX_USD_PER_RUN, cap "
            f"max_output_tokens, or use a cheaper alias."
        )

    last_exc: BaseException | None = None
    for attempt in range(max_retries + 1):
        try:
            return _call_provider(
                model_id,
                messages,
                max_output_tokens=max_output_tokens,
                temperature=temperature,
                timeout_s=timeout_s,
            )
        except Exception as exc:  # noqa: BLE001 — re-raised below
            last_exc = exc
            if attempt == max_retries or not _is_retryable(exc):
                raise
            # Exponential backoff with jitter: without jitter a fleet of workers
            # retries in lockstep and re-creates the spike that rate-limited it.
            base = 0.5 * (2**attempt)
            time.sleep(base + random.random() * base)

    assert last_exc is not None
    raise last_exc


def _call_provider(
    model_id: str,
    messages: list[Message],
    *,
    max_output_tokens: int | None,
    temperature: float | None,
    timeout_s: float,
) -> Completion:
    provider = spec(model_id).provider
    adapter = {"anthropic": _anthropic, "openai": _openai, "google": _google}[provider]
    with httpx.Client(timeout=timeout_s) as client:
        return adapter(client, model_id, messages, max_output_tokens, temperature)


def _require_key(value: str | None, name: str, provider: str) -> str:
    if not value:
        raise MissingCredentialError(
            f"{provider} needs {name}. Add it to .env (see .env.example) or "
            "pick a model from another provider."
        )
    return value


def _post(client: httpx.Client, url: str, **kwargs: Any) -> dict:
    res = client.post(url, **kwargs)
    res.raise_for_status()
    return res.json()


# ---- adapters -------------------------------------------------------------
# Each one does exactly two jobs: shape the request, and normalise the usage
# numbers. Everything else is handled above.


def _anthropic(client, model_id, messages, max_output_tokens, temperature) -> Completion:
    key = _require_key(settings.anthropic_api_key, "ANTHROPIC_API_KEY", "Anthropic")
    system = "\n\n".join(m.content for m in messages if m.role == "system")
    body: dict[str, Any] = {
        "model": model_id,
        "max_tokens": max_output_tokens or 4096,
        "messages": [
            {"role": m.role, "content": m.content} for m in messages if m.role != "system"
        ],
    }
    if system:
        body["system"] = system
    if temperature is not None:
        body["temperature"] = temperature

    raw = _post(
        client,
        "https://api.anthropic.com/v1/messages",
        headers={"x-api-key": key, "anthropic-version": "2023-06-01"},
        json=body,
    )
    u = raw["usage"]
    usage = Usage(
        input_tokens=u["input_tokens"],
        output_tokens=u["output_tokens"],
        cached_input_tokens=u.get("cache_read_input_tokens", 0) or 0,
    )
    text = "".join(b.get("text", "") for b in raw["content"] if b.get("type") == "text")
    return Completion(
        text=text,
        model=model_id,
        usage=usage,
        cost_usd=cost_of(
            model_id, usage.input_tokens, usage.output_tokens, usage.cached_input_tokens
        ),
        raw=raw,
    )


def _openai(client, model_id, messages, max_output_tokens, temperature) -> Completion:
    key = _require_key(settings.openai_api_key, "OPENAI_API_KEY", "OpenAI")
    body: dict[str, Any] = {
        "model": model_id,
        "messages": [{"role": m.role, "content": m.content} for m in messages],
    }
    if max_output_tokens:
        body["max_completion_tokens"] = max_output_tokens
    if temperature is not None:
        body["temperature"] = temperature

    raw = _post(
        client,
        "https://api.openai.com/v1/chat/completions",
        headers={"authorization": f"Bearer {key}"},
        json=body,
    )
    u = raw["usage"]
    usage = Usage(
        input_tokens=u["prompt_tokens"],
        output_tokens=u["completion_tokens"],
        cached_input_tokens=(u.get("prompt_tokens_details") or {}).get("cached_tokens", 0) or 0,
    )
    return Completion(
        text=raw["choices"][0]["message"].get("content") or "",
        model=model_id,
        usage=usage,
        cost_usd=cost_of(
            model_id, usage.input_tokens, usage.output_tokens, usage.cached_input_tokens
        ),
        raw=raw,
    )


def _google(client, model_id, messages, max_output_tokens, temperature) -> Completion:
    key = _require_key(settings.google_api_key, "GOOGLE_API_KEY", "Google")
    system = "\n\n".join(m.content for m in messages if m.role == "system")
    gen: dict[str, Any] = {}
    if max_output_tokens:
        gen["maxOutputTokens"] = max_output_tokens
    if temperature is not None:
        gen["temperature"] = temperature

    body: dict[str, Any] = {
        "contents": [
            {
                "role": "model" if m.role == "assistant" else "user",
                "parts": [{"text": m.content}],
            }
            for m in messages
            if m.role != "system"
        ],
        "generationConfig": gen,
    }
    if system:
        body["systemInstruction"] = {"parts": [{"text": system}]}

    raw = _post(
        client,
        f"https://generativelanguage.googleapis.com/v1beta/models/{model_id}:generateContent",
        headers={"x-goog-api-key": key},
        json=body,
    )
    u = raw["usageMetadata"]
    usage = Usage(
        input_tokens=u["promptTokenCount"],
        output_tokens=u.get("candidatesTokenCount", 0),
        cached_input_tokens=u.get("cachedContentTokenCount", 0) or 0,
    )
    parts = (raw.get("candidates") or [{}])[0].get("content", {}).get("parts", [])
    return Completion(
        text="".join(p.get("text", "") for p in parts),
        model=model_id,
        usage=usage,
        cost_usd=cost_of(
            model_id, usage.input_tokens, usage.output_tokens, usage.cached_input_tokens
        ),
        raw=raw,
    )
